samza-core/src/main/java/org/apache/samza/runtime/LocalApplicationRunner.java

```diff
@@ -119,24 +120,30 @@ public void run(StreamApplication app) {
         throw new SamzaException("No jobs to run.");
       }
       plan.getJobConfigs().forEach(jobConfig -> {
-          log.info("Starting job {} StreamProcessor with config {}", jobConfig.getName(), jobConfig);
-          String processorId = getProcessorId(config);
-          StreamProcessor processor =
-              createStreamProcessor(processorId, jobConfig, app, new LocalStreamProcessorListener(processorId));
-          processor.start();
+          log.debug("Starting job {} StreamProcessor with config {}", jobConfig.getName(), jobConfig);
+          LocalStreamProcessorLifeCycleListener listener = new LocalStreamProcessorLifeCycleListener();
+          StreamProcessor processor = createStreamProcessor(jobConfig, app, listener);
+          listener.setProcessor(processor);
           processors.add(processor);
+          processor.start();
         });
-      appStatus = ApplicationStatus.Running;
 
       // 4. block until the processors are done or there is a failure
-      awaitComplete();
+      latch.await();
+
+      if (throwable.get() != null) {
+        appStatus = ApplicationStatus.UnsuccessfulFinish;
+        throw throwable.get();
+      } else {
+        appStatus = ApplicationStatus.SuccessfulFinish;
+      }
 
     } catch (Throwable t) {
       appStatus = ApplicationStatus.UnsuccessfulFinish;
       throw new SamzaException("Failed to run application", t);
     } finally {
-      if (coordination != null) {
-        coordination.reset();
+      if (coordinationUtils != null) {
+        coordinationUtils.reset();
       }
     }
   }
```

Comment:
Same question as previous PR: Isn't kill supposed to be blocking in the current version? StreamProcessor stop is non-blocking.